File: frontend/src/engine/capabilities/groupSum.ts
/**
 * groupSum – Aggregate values by group using SUMIF or computed values.
 *
 * Strategy (preferFormula = true, default):
 *   Writes SUMIF/SUMIFS formulas so the results auto-update.
 *
 * Strategy (preferFormula = false):
 *   Reads data, computes group sums in JS, writes result values.
 */

import { CapabilityMeta, GroupSumParams, StepResult, ExecutionOptions } from "../types";
import { registry } from "../capabilityRegistry";
import { resolveRange } from "./rangeUtils";

const meta: CapabilityMeta = {
  action: "groupSum",
  description: "Sum values grouped by a column using SUMIF or computed aggregation",
  mutates: true,
  affectsFormatting: false,
};

async function handler(
  context: Excel.RequestContext,
  params: GroupSumParams,
  options: ExecutionOptions
): Promise<StepResult> {
  const { dataRange, groupByColumn, sumColumn, outputRange, preferFormula = true, includeHeaders } = params;

  if (options.dryRun) {
    return {
      stepId: "",
      status: "success",
      message: `Would compute grouped sums from ${dataRange}, output to ${outputRange}`,
    };
  }

  if (preferFormula) {
    return formulaGroupSum(context, params, options);
  }
  return computedGroupSum(context, params, options);
}

async function formulaGroupSum(
  context: Excel.RequestContext,
  params: GroupSumParams,
  options: ExecutionOptions
): Promise<StepResult> {
  options.onProgress?.("Reading data for group keys...");

  const dataRng = resolveRange(context, params.dataRange);
  dataRng.load("values");
  await context.sync();

  const data = dataRng.values;
  const startRow = params.includeHeaders ? 1 : 0;

  // Extract unique group keys
  const groupCol = params.groupByColumn - 1;
  const uniqueKeys: (string | number)[] = [];
  const seen = new Set<string>();
  for (let i = startRow; i < data.length; i++) {
    const key = String(data[i][groupCol]);
    if (!seen.has(key)) {
      seen.add(key);
      uniqueKeys.push(data[i][groupCol] as string | number);
    }
  }

  options.onProgress?.(`Found ${uniqueKeys.length} unique groups`);

  // Build output: group key column + SUMIF formula column
  const output: (string | number)[][] = [];
  if (params.includeHeaders) {
    output.push([String(data[0][groupCol]), `Sum of ${String(data[0][params.sumColumn - 1])}`]);
  }

  // Build column references for SUMIF
  const criteriaCol = getColumnFromRange(params.dataRange, groupCol);
  const sumCol = getColumnFromRange(params.dataRange, params.sumColumn - 1);

  for (const key of uniqueKeys) {
    const criteriaValue = typeof key === "string" ? `"${key}"` : key;
    output.push([key, `=SUMIF(${criteriaCol},${criteriaValue},${sumCol})` as any]);
  }

  const outRng = resolveRange(context, params.outputRange);
  // Write the group keys as values, formulas separately
  const values: (string | number | null)[][] = output.map((row) => [row[0], null]);
  const formulas: (string | null)[][] = output.map((row) => [null, String(row[1])]);

  // Write group keys
  const keyRange = outRng.getColumn(0).getResizedRange(output.length - 1, 0);
  keyRange.values = values.map((r) => [r[0]]);

  // Write SUMIF formulas
  const formulaRange = outRng.getColumn(1).getResizedRange(output.length - 1, 0);
  formulaRange.formulas = formulas.map((r) => [r[1] ?? ""]);

  await context.sync();

  return {
    stepId: "",
    status: "success",
    message: `Created ${uniqueKeys.length} SUMIF formulas in ${params.outputRange}`,
  };
}

async function computedGroupSum(
  context: Excel.RequestContext,
  params: GroupSumParams,
  options: ExecutionOptions
): Promise<StepResult> {
  options.onProgress?.("Reading data...");

  const dataRng = resolveRange(context, params.dataRange);
  dataRng.load("values");
  await context.sync();

  const data = dataRng.values;
  const startRow = params.includeHeaders ? 1 : 0;
  const groupCol = params.groupByColumn - 1;
  const sumCol = params.sumColumn - 1;

  options.onProgress?.("Computing group sums...");

  const groups = new Map<string, number>();
  for (let i = startRow; i < data.length; i++) {
    const key = String(data[i][groupCol]);
    const val = Number(data[i][sumCol]) || 0;
    groups.set(key, (groups.get(key) ?? 0) + val);
  }

  const output: (string | number)[][] = [];
  if (params.includeHeaders) {
    output.push([String(data[0][groupCol]), `Sum of ${String(data[0][sumCol])}`]);
  }
  for (const [key, sum] of groups) {
    output.push([key, sum]);
  }

  const outRng = resolveRange(context, params.outputRange);
  outRng.values = output;
  await context.sync();

  return {
    stepId: "",
    status: "success",
    message: `Computed ${groups.size} group sums, wrote to ${params.outputRange}`,
  };
}


function getColumnFromRange(rangeAddress: string, colOffset: number): string {
  const parts = rangeAddress.split("!");
  const ref = parts.length > 1 ? parts[1] : parts[0];
  const col = ref.match(/[A-Z]+/)?.[0] ?? "A";
  const offsetCol = offsetColumnLetter(col, colOffset);
  const prefix = parts.length > 1 ? parts[0] + "!" : "";
  return `${prefix}${offsetCol}:${offsetCol}`;
}

function offsetColumnLetter(col: string, offset: number): string {
  let num = 0;
  for (let i = 0; i < col.length; i++) {
    num = num * 26 + (col.charCodeAt(i) - 64);
  }
  num += offset;
  let result = "";
  while (num > 0) {
    const rem = (num - 1) % 26;
    result = String.fromCharCode(65 + rem) + result;
    num = Math.floor((num - 1) / 26);
  }
  return result || "A";
}

registry.register(meta, handler as any);
export { meta };

File: backend/app/execution/capabilities/group_sum.py
"""groupSum — aggregate values by group using SUMIF formulas or computed values.

Two strategies (selected by ``preferFormula`` — default True):
  * Formula strategy: write SUMIF formulas that auto-update when source data
    changes. Group keys are written as values, the sum column as formulas.
  * Computed strategy: read data, build a {group: sum} dict in Python, write
    the fully materialized 2-column table.

When ``includeHeaders`` is true (default), the first row of the source range
is treated as headers and an ``Sum of <name>`` header is written to the output.
"""

from __future__ import annotations

import re
from typing import Any

from app.execution.base import ExecutorContext
from app.execution.capability_registry import registry
from app.execution.range_utils import normalize_token, resolve_range


_SHEET_SAFE_RE = re.compile(r"^[A-Za-z_][A-Za-z0-9_.]*$")


def _strip_workbook_qualifier(address: str) -> str:
    """Drop the leading ``[Book.xlsx]`` segment if present."""
    addr = normalize_token(address)
    return re.sub(r"^\[.*?\]", "", addr)


def _quote_sheet_in_ref(ref: str) -> str:
    """Mirror the TS ``quoteSheetInRef`` helper used inside formula strings."""
    bang = ref.rfind("!")
    if bang == -1:
        return ref
    sheet_part = ref[:bang]
    cell_part = ref[bang + 1 :]
    if sheet_part.startswith("'") and sheet_part.endswith("'"):
        return ref
    if _SHEET_SAFE_RE.match(sheet_part):
        return ref
    escaped = sheet_part.replace("'", "''")
    return f"'{escaped}'!{cell_part}"


def _offset_column_letter(col: str, offset: int) -> str:
    num = 0
    for ch in col:
        num = num * 26 + (ord(ch.upper()) - 64)
    num += offset
    result = ""
    while num > 0:
        num, rem = divmod(num - 1, 26)
        result = chr(65 + rem) + result
    return result or "A"


def _column_for_range(range_address: str, col_offset: int) -> str:
    """Return a sheet-qualified ``A:A`` column reference for SUMIF."""
    parts = range_address.split("!")
    ref = parts[1] if len(parts) > 1 else parts[0]
    m = re.search(r"[A-Z]+", ref, re.IGNORECASE)
    col = m.group(0).upper() if m else "A"
    offset_col = _offset_column_letter(col, col_offset)
    prefix = parts[0] + "!" if len(parts) > 1 else ""
    return _quote_sheet_in_ref(f"{prefix}{offset_col}:{offset_col}")


def _normalize_to_2d(raw: Any, shape: tuple[int, int]) -> list[list[Any]]:
    if raw is None:
        return []
    if not isinstance(raw, list):
        return [[raw]]
    if raw and not isinstance(raw[0], list):
        rows_, _ = shape
        return [list(raw)] if rows_ == 1 else [[v] for v in raw]
    return [list(r) for r in raw]


def _to_float(x: Any) -> float:
    try:
        if x is None or x == "":
            return 0.0
        return float(x)
    except (TypeError, ValueError):
        return 0.0


def _formula_group_sum(ctx: ExecutorContext, params: dict[str, Any]) -> dict[str, Any]:
    data_range = params["dataRange"]
    group_by = int(params["groupByColumn"])
    sum_col = int(params["sumColumn"])
    output_range = params["outputRange"]
    include_headers = bool(params.get("includeHeaders", True))

    data_rng = resolve_range(ctx.workbook_handle, data_range)
    data = _normalize_to_2d(data_rng.value, data_rng.shape)

    start_row = 1 if include_headers else 0
    group_idx = group_by - 1

    # Extract unique group keys (preserve first-seen ordering).
    unique_keys: list[Any] = []
    seen: set[str] = set()
    for i in range(start_row, len(data)):
        if group_idx >= len(data[i]):
            continue
        key_display = data[i][group_idx]
        key_str = str(key_display)
        if key_str in seen:
            continue
        seen.add(key_str)
        unique_keys.append(key_display)

    data_range_for_formula = _strip_workbook_qualifier(data_range)
    criteria_col = _column_for_range(data_range_for_formula, group_idx)
    sum_col_ref = _column_for_range(data_range_for_formula, sum_col - 1)

    # Write optional header row.
    out_rng = resolve_range(ctx.workbook_handle, output_range)
    sheet = out_rng.sheet
    start_out_row = out_rng.row
    start_out_col = out_rng.column

    def _col_letter(col_num: int) -> str:
        letters = ""
        n = col_num
        while n > 0:
            n, rem = divmod(n - 1, 26)
            letters = chr(65 + rem) + letters
        return letters

    key_col_letter = _col_letter(start_out_col)
    val_col_letter = _col_letter(start_out_col + 1)

    row_cursor = start_out_row
    if include_headers and data:
        header_group = str(data[0][group_idx]) if group_idx < len(data[0]) else ""
        header_sum = str(data[0][sum_col - 1]) if (sum_col - 1) < len(data[0]) else ""
        sheet.range(f"{key_col_letter}{row_cursor}").value = header_group
        sheet.range(f"{val_col_letter}{row_cursor}").value = f"Sum of {header_sum}"
        row_cursor += 1

    # Write group keys as values, SUMIF formulas next to them.
    if unique_keys:
        keys_block = [[k] for k in unique_keys]
        formulas_block = [
            [
                f"=SUMIF({criteria_col},"
                + (f'"{k}"' if isinstance(k, str) else str(k))
                + f",{sum_col_ref})"
            ]
            for k in unique_keys
        ]
        n = len(unique_keys)
        sheet.range(f"{key_col_letter}{row_cursor}").resize(n, 1).value = keys_block
        sheet.range(f"{val_col_letter}{row_cursor}").resize(n, 1).formula = formulas_block

    return {
        "status": "success",
        "message": f"Created {len(unique_keys)} SUMIF formulas in {output_range}.",
        "outputs": {"outputRange": output_range, "groupCount": len(unique_keys)},
    }


def _computed_group_sum(ctx: ExecutorContext, params: dict[str, Any]) -> dict[str, Any]:
    data_range = params["dataRange"]
    group_by = int(params["groupByColumn"])
    sum_col = int(params["sumColumn"])
    output_range = params["outputRange"]
    include_headers = bool(params.get("includeHeaders", True))

    data_rng = resolve_range(ctx.workbook_handle, data_range)
    data = _normalize_to_2d(data_rng.value, data_rng.shape)

    start_row = 1 if include_headers else 0
    group_idx = group_by - 1
    sum_idx = sum_col - 1

    groups: dict[str, list[Any]] = {}
    for i in range(start_row, len(data)):
        if group_idx >= len(data[i]):
            continue
        display = data[i][group_idx]
        key = str(display)
        if key not in groups:
            groups[key] = [display, 0.0]
        value = _to_float(data[i][sum_idx] if sum_idx < len(data[i]) else 0)
        groups[key][1] += value

    output: list[list[Any]] = []
    if include_headers and data:
        header_group = str(data[0][group_idx]) if group_idx < len(data[0]) else ""
        header_sum = str(data[0][sum_idx]) if sum_idx < len(data[0]) else ""
        output.append([header_group, f"Sum of {header_sum}"])
    for display, total in groups.values():
        output.append([display, total])

    out_rng = resolve_range(ctx.workbook_handle, output_range)
    rows = len(output)
    cols = 2 if output else 0
    if rows and cols:
        out_rng.resize(rows, cols).value = output

    return {
        "status": "success",
        "message": f"Computed {len(groups)} group sums, wrote to {output_range}.",
        "outputs": {"outputRange": output_range, "groupCount": len(groups)},
    }


def handler(ctx: ExecutorContext, params: dict[str, Any]) -> dict[str, Any]:
    required = ("dataRange", "groupByColumn", "sumColumn", "outputRange")
    missing = [k for k in required if params.get(k) in (None, "")]
    if missing:
        return {
            "status": "error",
            "message": f"groupSum missing parameters: {', '.join(missing)}.",
        }

    if ctx.dry_run:
        return {
            "status": "preview",
            "message": (
                f"Would compute grouped sums from {params['dataRange']}, "
                f"output to {params['outputRange']}."
            ),
        }

    prefer_formula = bool(params.get("preferFormula", True))

    try:
        if prefer_formula:
            return _formula_group_sum(ctx, params)
        return _computed_group_sum(ctx, params)
    except Exception as exc:  # noqa: BLE001
        return {"status": "error", "message": f"groupSum failed: {exc}", "error": str(exc)}


registry.register("groupSum", handler, mutates=True)
